torch/csrc/jit/passes/memory_planning.h

```diff
@@ -0,0 +1,151 @@
+#pragma once
+
+#include <torch/csrc/jit/ir/ir.h>
+#include <torch/csrc/jit/runtime/static/impl.h>
+
+namespace torch {
+namespace jit {
+enum class Strategy {
+  NAIVE = 0,
+};
+
+inline const char* toString(Strategy s) {
+  switch (s) {
+    case Strategy::NAIVE:
+      return "NAIVE";
+    default:
+      return "UNKNOWN STRATEGY";
+  }
+}
+
+inline std::ostream& operator<<(std::ostream& str, Strategy rhs) {
+  return str << toString(rhs);
+}
+
+typedef struct MemRegion {
+  size_t offset;
+  size_t size;
+} MemRegion;
+
+inline std::ostream& operator<<(std::ostream& str, MemRegion reg) {
+  return str << "{offset: " << reg.offset << ", size: " << reg.size << "}";
+}
+
+inline bool operator==(const MemRegion& lhs, const MemRegion& rhs) {
+  return lhs.offset == rhs.offset && lhs.size == rhs.size;
+}
+
+struct regionSizeCmp {
+  bool operator()(const MemRegion& reg1, const MemRegion& reg2) const {
+    return reg1.size == reg2.size ? reg1.offset < reg2.offset
+                                  : reg1.size < reg2.size;
+  }
+};
+
+struct regionOffsetCmp {
+  bool operator()(const MemRegion& reg1, const MemRegion& reg2) const {
+    return reg1.offset == reg2.offset ? reg1.size < reg2.size
+                                      : reg1.offset < reg2.offset;
+  }
+};
+
+bool overlapMemRegion(const MemRegion& reg1, const MemRegion& reg2);
+
+struct UniqueLiveRange {
+  LiveRange lvr;
+  std::string id;
+};
+
+bool overlapLiveRange(
+    const UniqueLiveRange& ulvr1,
+    const UniqueLiveRange& ulvr2);
+
+inline std::ostream& operator<<(std::ostream& str, UniqueLiveRange rhs) {
+  return str << "{id: " << rhs.id << ", lvr: " << rhs.lvr << "}";
+}
+
+inline bool operator==(const UniqueLiveRange lhs, const UniqueLiveRange rhs) {
+  return lhs.lvr == rhs.lvr && lhs.id == rhs.id;
+}
+
+struct liveRangeStartCmp {
+  bool operator()(const UniqueLiveRange& u1, const UniqueLiveRange& u2) const {
+    return u1.lvr.begin == u2.lvr.begin
+        ? (u1.lvr.end == u2.lvr.end ? u1.id < u2.id : u1.lvr.end < u2.lvr.end)
+        : u1.lvr.begin < u2.lvr.begin;
+  }
+};
+
+struct liveRangeEndCmp {
+  bool operator()(const UniqueLiveRange& u1, const UniqueLiveRange& u2) const {
+    return u1.lvr.end == u2.lvr.end
+        ? (u1.lvr.begin == u2.lvr.begin ? u1.id < u2.id
+                                        : u1.lvr.begin < u2.lvr.begin)
+        : u1.lvr.end < u2.lvr.end;
+  }
+};
+
+template <typename Value>
+using SortedLiveRangeMap = std::map<UniqueLiveRange, Value, liveRangeStartCmp>;
+struct TORCH_API MemAllocation {
+  UniqueLiveRange ulvr;
+  MemRegion reg;
+};
+
+inline std::ostream& operator<<(std::ostream& str, MemAllocation rhs) {
+  return str << rhs.ulvr << ", " << rhs.reg;
+}
+
+inline bool operator==(const MemAllocation lhs, const MemAllocation rhs) {
+  return lhs.ulvr == rhs.ulvr && lhs.reg == rhs.reg;
+}
+
+inline bool valid_add(size_t a, size_t b) {
```

Comment:
why do we need this ? seems like a good sign we should be using `int64_t` instead

Reply:
not sure i follow - `int64_t` would make the problem worse - 1 less bit and much harder to predict overflow that as well lead to UB (whereas unsigned types only roll over)

Reply:
What are we doing that requires subtracting size_t from each other and to more than 32 bits of precision ?

Reply:
oop thats above in the code...